firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/metadata/MetaDataStoreTest.java

```diff
@@ -181,28 +186,37 @@ public void testSetSameKeysRaceCondition_preserveLastEntryValue() throws Excepti
                 put(KEY_2, "20000");
               }
             };
-    UserMetadata userMetadata = new UserMetadata(SESSION_ID_1, fileStore, backgroundExecutor);
-    for (int index = 0; index <= 10000; index++) {
-      userMetadata.setCustomKey(KEY_1, String.valueOf(index));
-      userMetadata.setCustomKey(KEY_2, String.valueOf(index * 2));
+    UserMetadata userMetadata = new UserMetadata(SESSION_ID_1, fileStore, userActionsExecutor);
+    for (int times = 0; times <= 100;  times++) {
+      for (int index = 0; index <= 10000; index++) {
+        userMetadata.setCustomKey(KEY_1, String.valueOf(index));
+        userMetadata.setCustomKey(KEY_2, String.valueOf(index * 2));
+        if (index == 0 || index == 10000) {
+          Logger.getLogger().i("Finish write the key data: " + index + " time: " + times);
+        }
+      }
+
+      int time = times;
+      userMetadata.backgroundSequentialExecutor.execute(() -> {
+        Logger.getLogger().i("Start read the key data, times: "  + time );
+        final Map<String, String> readKeys = storeUnderTest.readKeyData(SESSION_ID_1);
+        Logger.getLogger().i("Data read: " + readKeys);
+        assertThat(readKeys.get(KEY_1)).isEqualTo("10000");
+        assertThat(readKeys.get(KEY_2)).isEqualTo("20000");
+        assertEqualMaps(keys, readKeys);
+      });
+      Thread.sleep(40);
```

Comment:
Need this sleep otherwise read might happen after the next write.(read can be delayed and it is not thread safe)